Source: Customer
Description:
  Thinko reported in arm_adjust_cost.

diff -Nurp gcc-3.3/gcc/config/arm.orig/arm.c gcc-3.3/gcc/config/arm/arm.c
--- gcc-3.3/gcc/config/arm.orig/arm.c	2003-09-03 12:55:47.000000000 -0400
+++ gcc-3.3/gcc/config/arm/arm.c	2003-09-03 15:46:51.000000000 -0400
@@ -3039,8 +3039,8 @@ arm_adjust_cost (insn, link, dep, cost)
      on precisely how certain input operands are used.  */
   if (arm_is_xscale
       && REG_NOTE_KIND (link) == 0
-      && recog_memoized (insn) < 0
-      && recog_memoized (dep) < 0)
+      && recog_memoized (insn) >= 0
+      && recog_memoized (dep) >= 0)
     {
       int shift_opnum = get_attr_shift (insn);
       enum attr_type attr_type = get_attr_type (dep);
